### PR TITLE
[WIP] Patch dcgan

### PR DESCRIPTION
**Pull request by Koukyosyumei**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

I would like to solve #418. I quickly implemented two solutions suggested in the [forum](https://discuss.pytorch.org/t/double-charging-privacy-in-dcgan-example/149098). Though I have created two separate py scripts for the solutions, I will experiment with the model performance of each solution and merge the better one. Since this is the first PR for opacus, any suggestions and help are welcome!

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [ ] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by Koukyosyumei** [comment]
I have run both scripts and found that only patch_1 (clipping gradients for both fake and real data) works. I think it is reasonable that the second approach, which uses two optimizers, makes the gradients for actual data too small compared to real data due to the gradient clipping. Thus, I suggest using the first solution.

**Comment by romovpa** [comment]
@Koukyosyumei Thank you very much for working on this! 
May I ask you to provide some quantified results for the run that can show that the new version works correctly? You can put them in the testing section of this PR.

**Comment by Koukyosyumei** [comment]
@romovpa 

I am sorry for the late reply. 

## 1. Loss

The losses of the original implementation, solution 1, and solution 2 correspond to green, yellow, and blur lines. Due to the memory limitation, solution 2  works only batch size = 32for my environment (google colab).

batchsize_32
batchsize_64

## 2. Examples of generated images

### batch size 64

- original implementation

fake_samples_epoch_022

- solution 1

fake_samples_epoch_021

### batch size 32

- original implementation

- solution 1

- solution 2


## 3. Final $\epsilon$

batch size 64

|    |  $\epsilon$   |
|---|---|
|original| 4.82 |
|solution 1| 3.36|
